src/helpers/block_request.rs

```diff
@@ -0,0 +1,124 @@
+// Copyright (C) 2019-2021 Aleo Systems Inc.
+// This file is part of the snarkOS library.
+
+// The snarkOS library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkOS library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
+
+use snarkvm::dpc::prelude::*;
+
+use std::hash::{Hash, Hasher};
+
+///
+/// A request for a block with the specified height and an optional hash.
+///
+#[derive(Clone, Debug)]
+pub struct BlockRequest<N: Network> {
+    block_height: u32,
+    block_hash: Option<N::BlockHash>,
+}
+
+impl<N: Network> BlockRequest<N> {
+    /// Returns the block height stored in the request.
+    pub fn block_height(&self) -> u32 {
+        self.block_height
+    }
+
+    /// Returns the block hash stored in the request, if it exists.
+    pub fn block_hash(&self) -> Option<N::BlockHash> {
+        self.block_hash
+    }
+}
+
+impl<N: Network> From<u32> for BlockRequest<N> {
+    fn from(height: u32) -> Self {
+        Self {
+            block_height: height,
+            block_hash: None,
+        }
+    }
+}
+
+impl<N: Network> From<(u32, Option<N::BlockHash>)> for BlockRequest<N> {
+    fn from((height, hash): (u32, Option<N::BlockHash>)) -> Self {
+        Self {
+            block_height: height,
+            block_hash: hash,
+        }
+    }
+}
+
+// The height is the primary key, so use only it for hashing purposes.
+impl<N: Network> PartialEq for BlockRequest<N> {
+    fn eq(&self, other: &Self) -> bool {
+        self.block_height == other.block_height
+    }
+}
+
+impl<N: Network> Eq for BlockRequest<N> {}
+
+// The k1 == k2 -> hash(k1) == hash(k2) rule must hold.
+impl<N: Network> Hash for BlockRequest<N> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.block_height.hash(state);
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use snarkvm::{dpc::testnet2::Testnet2, prelude::UniformRand};
+
+    use rand::{thread_rng, Rng};
+
+    const ITERATIONS: usize = 1_000_000;
```

Comment:
I don't see any harm in iterating (beyond making the tests slower) but is there any reason you suspect different values would change the behaviour in this case?

Reply:
For this many iterations I'd rather use `thread_rng` to seed a different, faster source of randomness to use in the loop; it's something I'd like to do for other tests too, including snarkVM ones.

Reply:
We can do that in one go to keep it consistent in design.

The rationale for multiple iterations is more to ensure the Default::default() block hash Eq case is accounted for.

I put 1 million iterations arbitrarily, it could be 1000 as well.

Reply:
I dropped the iteration count to 5.